pkg/resources/user_password_policy_attachment.go

```diff
@@ -0,0 +1,169 @@
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	ctx := context.Background()
+
+	userName := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(d.Get("user_name").(string))
+	passwordPolicyDatabase := sdk.NewAccountIdentifierFromFullyQualifiedName(d.Get("password_policy_database").(string))
+	passwordPolicySchema := sdk.NewAccountIdentifierFromFullyQualifiedName(d.Get("password_policy_schema").(string))
+	passwordPolicyName := sdk.NewAccountIdentifierFromFullyQualifiedName(d.Get("password_policy_name").(string))
+	passwordPolicy := sdk.NewSchemaObjectIdentifier(passwordPolicyDatabase.Name(), passwordPolicySchema.Name(), passwordPolicyName.Name())
+
+	err := client.Users.Alter(ctx, userName, &sdk.AlterUserOptions{
+		Set: &sdk.UserSet{
+			PasswordPolicy: passwordPolicy,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_database", passwordPolicyDatabase.Name()); err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_schema", passwordPolicySchema.Name()); err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_name", passwordPolicyName.Name()); err != nil {
+		return err
+	}
+	if err := d.Set("user_name", helpers.EncodeSnowflakeID(userName)); err != nil {
+		return err
+	}
+	d.SetId(fmt.Sprintf(`%s|%s`, helpers.EncodeSnowflakeID(passwordPolicy), helpers.EncodeSnowflakeID(userName)))
+
+	return ReadUserPasswordPolicyAttachment(d, meta)
+}
+
+func ReadUserPasswordPolicyAttachment(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	ctx := context.Background()
+	userName := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(d.Get("user_name").(string))
+	policyReferences, err := client.PolicyReferences.GetForEntity(ctx, &sdk.GetForEntityPolicyReferenceRequest{
+		// Note: I cannot insert both single and double quotes in the SDK, so for now I need to do this
+		RefEntityName:   sdk.String(userName.FullyQualifiedName()),
+		RefEntityDomain: sdk.String("user"),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Note: this should never happen, but just in case: so far, Snowflake only allows one Password Policy per user.
+	if len(policyReferences) > 1 {
+		return fmt.Errorf("internal error: multiple policy references attached to a user. This should never happen")
+	}
+
+	// Note: this means the resource has been deleted outside of Terraform.
+	if len(policyReferences) == 0 {
+		d.SetId("")
+		return nil
+	}
+	if err := d.Set("password_policy_database", sdk.NewAccountIdentifierFromFullyQualifiedName(policyReferences[0].PolicyDb).Name()); err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_schema", sdk.NewAccountIdentifierFromFullyQualifiedName(policyReferences[0].PolicySchema).Name()); err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_name", sdk.NewAccountIdentifierFromFullyQualifiedName(policyReferences[0].PolicyName).Name()); err != nil {
+		return err
+	}
```

Comment:
here you should add `d.Set("user_name")` to detect Snowflake changes. This information could be probably fetched with https://docs.snowflake.com/en/user-guide/admin-user-management#considerations (Tracking password policy usage). It's not in the client, so:
- leave the comment and we'll do it
- use raw SQL (client.UsafeExec or something like that) in place to get the user_name and we'll move it to the client later 
- or you can add the `ReferencesFunctions` interface to the client struct (next to the system-defined functions) and use the raw SQL there to retrieve the needed username. I guess `ConvertionFunctions` are pretty good reference.
```go
// client.go
type Client struct {
	config         *gosnowflake.Config
	db             *sqlx.DB
	sessionID      string
	accountLocator string
	dryRun         bool
	traceLogs      []string

	// System-Defined Functions
	ContextFunctions     ContextFunctions
        <here>
        // ...
}

Reply:
Oh ok, I saw you already did it, so you can use the function here to get the username

Reply:
Ok, I see. We can probably leave it as is then. cc: @sfc-gh-asawicki